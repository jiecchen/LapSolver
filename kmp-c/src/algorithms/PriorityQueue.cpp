#include <utility>
#include <algorithm>
#include <set>

#include "PriorityQueue.h"

using namespace std;

PriorityQueue::PriorityQueue (int capacity) {
    heap = new int[capacity];
    heap_index = new int[capacity];
    priority = new double[capacity];

    heap_index[0:capacity] = -1;

    size = 0;
}

PriorityQueue::~PriorityQueue () {
    delete[] heap;
    delete[] heap_index;
    delete[] priority;
}

void PriorityQueue::Push (int key, double pri) {
    heap[size] = key;
    heap_index[key] = size;
    priority[key] = pri;
    size++;

    UpHeap(key);
}

void PriorityQueue::DecreaseKey (int key, double pri) {
    priority[key] = pri;
    UpHeap(key);
    DownHeap(key);
}

pair<int, double> PriorityQueue::Pop () {
    pair<int, double> ret(heap[0], priority[heap[0]]);

    SwapKeys(0, size-1);
    heap_index[heap[size-1]] = -1;
    size--;

    DownHeap(heap[0]);

    return ret;
}

void PriorityQueue::UpHeap (int key) {
    if (heap_index[key] == 0) return;

    int parent = (heap_index[key]-1) / 2;
    if (priority[heap[parent]] > priority[key]) {
        SwapKeys (key, heap[parent]);
        UpHeap (key);
    }
}

void PriorityQueue::DownHeap (int key) {
    int left_child = 2*(heap_index[key]+1)-1, right_child = 2*(heap_index[key]+1);
    int to_swap = key;

    if (left_child < size && priority[heap[left_child]] < priority[heap[to_swap]]) {
        to_swap = left_child;
    }

    if (right_child < size && priority[heap[right_child]] < priority[heap[to_swap]]) {
        to_swap = right_child;
    }

    if (to_swap != key) {
        SwapKeys (key, heap[to_swap]);
        DownHeap (key);
    }
}

void PriorityQueue::SwapKeys (int key1, int key2) {
    swap(heap[key1], heap[key2]);
    swap(priority[heap[key1]], priority[heap[key2]]);
    swap(heap_index[heap[key1]], heap_index[heap[key2]]);
}
