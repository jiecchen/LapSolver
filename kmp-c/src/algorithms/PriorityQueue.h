#pragma once

#include <utility>

class PriorityQueue {
  public:
    PriorityQueue (int capacity);
    ~PriorityQueue ();

    void Push (int key, double pri);
    void DecreaseKey (int key, double pri);
    std::pair<int, double> Pop ();

  private:
    void UpHeap (int key);
    void DownHeap (int key);
    void SwapKeys (int key1, int key2);

    int* heap;
    int* heap_index;
    double* priority;
    int size;
};
