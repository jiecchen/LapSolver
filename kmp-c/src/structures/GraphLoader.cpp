#include "GraphLoader.h"
#include <fstream>
#include <sstream>
#include <string>
using namespace std;

Graph GraphLoader::fromArrays(aligned_vector<int> u, aligned_vector<int> v, aligned_vector<double> w)
{
    return Graph(EdgeList(u, v, w));
}

Graph GraphLoader::fromFile(const string &filename)
{
    aligned_vector<int> u = aligned_vector<int>();
    aligned_vector<int> v = aligned_vector<int>();
    aligned_vector<double> w = aligned_vector<double>();

    string line;
    ifstream ijvFile(filename);
    if (!ijvFile)
        throw -1;

    while (getline(ijvFile, line))
    {
        if (line.empty()) break;
        istringstream iss(line);
        int a, b;
        double c;
        if (!(iss >> a >> b >> c))
            throw u.size();
        u.push_back(a);
        v.push_back(b);
        w.push_back(c);
    }

    return fromArrays(u, v, w);
}