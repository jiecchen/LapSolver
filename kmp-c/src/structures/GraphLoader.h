#pragma once

#include <string>
#include <iostream>
#include "structures/Graph.h"
#include "util/aligned.h"

class GraphLoader
{
public:
	
	static Graph fromArrays(aligned_vector<int> u, aligned_vector<int> v, aligned_vector<double> w);
	static Graph fromFile(const std::string &filename);
    static Graph fromStream(std::istream &inStream);
    static Graph fromStdin();

};