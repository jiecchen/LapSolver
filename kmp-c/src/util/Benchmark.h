#pragma once
#include <string>
#include <chrono>
#include <stdexcept>
#include <cstdio>

template<typename Func>
class Benchmark
{
public:
    Benchmark(int argc, char **argv, Func f)
        : name(argv[0])
    {
        for (int i = 1; i < argc; ++i)
        {
            std::string flag(argv[i]);
            if (flag == "-v")
                echo = true;
            else if (flag == "-n" && i + 1 < argc)
            {
                try
                {
                    nTrials = std::stoi(argv[i + 1]);
                }
                catch (...)
                {
                    fprintf(stderr, "bad value for flag -n '%s'\n", argv[i + 1]);
                    break;
                }
            }
        }
        runBenchmark(f);
    }

    Benchmark(std::string name, Func f, bool echo = false, int nTrials = 1)
        : name(name), echo(echo), nTrials(nTrials)
    {
        runBenchmark(f);
    }

    ~Benchmark()
    {
        if (echo)
            fprintf(stderr, "%s completed in %dms (on average over %d trials)\n", name.c_str(), avgTime.count(), nTrials);
    }

    void runBenchmark(Func f)
    {
        auto start = std::chrono::high_resolution_clock::now();
        for (int i = 0; i < nTrials; ++i)
            f();
        auto stop = std::chrono::high_resolution_clock::now();
        avgTime = std::chrono::duration_cast<std::chrono::milliseconds>(stop - start) / nTrials;
    }

private:
    std::string name;
    bool echo = false;
    int nTrials = 1;
    std::chrono::milliseconds avgTime;
};

template<typename Func>
Benchmark<Func> make_benchmark(int argc, char **argv, Func f)
{
    return Benchmark<Func>(argc, argv, f);
}
