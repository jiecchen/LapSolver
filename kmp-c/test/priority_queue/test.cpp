// Test for Priority Queue

#include <cstdlib>
#include <string>
#include <memory>
#include <structures/GraphLoader.h>
#include <algorithms/PriorityQueue.h>
#include <util/Benchmark.h>

using namespace std;

int main(int argc, char *argv[])
{
    atexit(mkl_free_buffers);

    int N, M;
    cin >> N >> M;

    std::shared_ptr<PriorityQueue> pq;

    auto bench = make_benchmark(argc, argv, [&] () {
        pq.reset(new PriorityQueue(N));

        for (int i = 0; i < M; i++) {
            string OP;
            cin >> OP;

            if (OP == "PUSH") {
                int x;
                double v;
                cin >> x >> v;

                pq->Push(x,v);
            }

            if (OP == "POP") {
                pair <int, double> p;
                p = pq->Pop();

                cout << p.first << " " << p.second << endl;
            }

            if (OP == "DEC") {
                int x;
                double v;
                cin >> x >> v;

                pq->DecreaseKey(x,v);    
            }
        }
    });

    

    return 0;
}
