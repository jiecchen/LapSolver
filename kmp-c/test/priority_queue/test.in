10 18
PUSH 1 30
PUSH 5 14
PUSH 0 29
POP
PUSH 5 194
PUSH 3 33
PUSH 2 59
POP
DEC 5 1
POP
PUSH 7 10
PUSH 9 23
DEC 9 21
DEC 9 17
DEC 9 1
POP
POP
POP