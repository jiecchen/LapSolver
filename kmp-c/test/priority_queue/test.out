5 14
0 29
5 1
9 1
7 10
1 30